graylog2-server/src/main/java/org/graylog/security/certutil/CaService.java

```diff
@@ -0,0 +1,138 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+@Singleton
+public class CaService {
+    private static final Logger LOG = LoggerFactory.getLogger(CaService.class);
+    // TODO: clarify default value
+    private static int DEFAULT_VALIDITY = 10 * 365;
+    private final KeystoreFileStorage keystoreFileStorage;
+    private final KeystoreMongoStorage keystoreMongoStorage;
+    private final NodeId nodeId;
+    private final CACreator caCreator;
+    private final CaConfiguration configuration;
+
+    private Optional<CA> currentCA = Optional.empty();
+
+    private Optional<String> password;
+
+    @Inject
+    public CaService(final Configuration configuration,
+                     final KeystoreFileStorage keystoreFileStorage,
+                     final KeystoreMongoStorage keystoreMongoStorage,
+                     final NodeId nodeId,
+                     final CACreator caCreator) {
+        this.keystoreFileStorage = keystoreFileStorage;
+        this.keystoreMongoStorage = keystoreMongoStorage;
+        this.nodeId = nodeId;
+        this.caCreator = caCreator;
+        this.configuration = configuration;
+
+        if(configuration.getCaKeystoreFile() != null && Files.exists(configuration.getCaKeystoreFile())) {
+            currentCA = Optional.of(new CA("local CA", CAType.LOCAL));
+        }
+        password = Optional.ofNullable(configuration.getCaPassword());
+    }
+
+    public CA get() {
+        return currentCA.get();
+    }
+
+    // TODO: write local CA and password / also password for generated/uploaded CA into MongoDB
+
+    public CA create(final String password, final Integer daysValid) throws CACreationException {
+        // a given password overrides an eventually configured password from the config
+        if(password != null) {
+            this.password = Optional.of(password);
+        }
+
+        try {
+            final var pass = this.password.map(String::toCharArray).orElse(null);
+            final Duration certificateValidity = Duration.ofDays(daysValid == null || daysValid == 0 ? DEFAULT_VALIDITY: daysValid);
+            KeyStore keyStore = caCreator.createCA(pass, certificateValidity);
+            keystoreMongoStorage.writeKeyStore(nodeId, keyStore, pass);
+        } catch (Exception ex) {
+            LOG.error("Could not generate CA: " + ex.getMessage(), ex);
+            throw new CACreationException("Could not generate CA: " + ex.getMessage(), ex);
+        }
+
+        currentCA = Optional.of(new CA("generated CA", CAType.GENERATED));
+        return currentCA.get();
```

Comment:
There is not need to create an Optional and ... fetch its value right away.

Reply:
I clarified by adding a `this.` because it's actually saved into the singleton first. but I guess it should get refactored to become stateless